Updated to the RC3 standard of JSON API

eneuhauser opened the pull request:
Per [JSON API RC3 release](https://github.com/json-api/json-api/pull/483), the spec is close to being finalized. Now there is a line in the sand, it's easy to build against the RC3 standard until either another release candidate or 1.0 is ratified. The main feature that this pull request supports that #68 does not is the consistent linkage change, where a linked resource will have a `linkage` property that includes the `type` and `id`. The serializer is not backwards compatible and only works with the RC3 standard.

One concept of JSON API I have not been able to support is the `self` link. The current interface for the adapter does not allow accessing the *owner* record when doing a `createRecord` or `updateRecord`. I'm storing the `self` link within the resource's links, but there currently isn't a way to access it.

On the current project I'm working on, I have updated my server to be compliant with the RC3 standard and this adapter and serializer is working well.

green-arrow commented:
The following could be because I'm not following the spec correctly, but humor me. In a case where there are `links` in objects in the `included` array, an error is thrown if the link is not a string, but an object. This means that I cannot, in an object in the `included` array, provide `linkage` to other object in the `included` array.

Example:

```json
{
  "data": {
    "type": "myType",
    "id": "1",
    "name": "myType Name",
    "links": {
      "self": "/myTypes/1",
      "linkOne": {
        "self": "/myTypes/1/linkOne",
        "linkage": { "type": "subLink", "id": "1" }
      }
    }
  },
  "included": [{
    "type": "subLink",
    "id": "1",
    "name": "subLink Name",
    "links": {
      "self": "/myTypes/1/linkOne",
      "linkTwo": {
        "self": "/myTypes/1/linkOne/linkTwo",
        "linkage": { "type": "otherSubLink", "id": "1" } // <---- Link to another sideloaded record
      }
    }
  }, {
    "type": "otherSubLink",
    "id": "1",
    "name": "otherSubLink Name",
    "links": {
      "self": "/myTypes/1/linkOne/linkTwo"
    }
  }]
}
```

Thoughts?

eneuhauser commented:
Hi @green-arrow,

I believe your interpretation is correct and I just committed a change that should fix this issue. I also included additional tests to check for this scenario. Thanks for the feedback.

green-arrow commented:
@eneuhauser I just pulled down your branch with the recent update and it's working beautifully. Tests look good as well.

@kurko - is there anything in particular that is stopping this from being merged?

kurko commented:
No. Will review this afternoon!
On Thu, Mar 19, 2015 at 8:52 AM Andrew Walton <notifications@github.com>
wrote:

> @eneuhauser <https://github.com/eneuhauser> I just pulled down your
> branch with the recent update and it's working beautifully. Tests look good
> as well.
>
> @kurko <https://github.com/kurko> - is there anything in particular that
> is stopping this from being merged?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kurko/ember-json-api/pull/71#issuecomment-83522320>.
>

jonkoops commented:
I'm still missing the right Content-Type header for exchanging data. From the specification:
> JSON API requires use of the JSON API media type (application/vnd.api+json) for exchanging data.

I suggest overwriting the `ajaxOptions` method in the adapter to change the content type to the right value, like so:

```js
ajaxOptions: function(url, type, options) {
  var hash = this._super(url, type, options);
  hash.contentType = 'application/vnd.api+json';
  return hash;
}
```

green-arrow commented:
Additional comment in regards to `findBelongsTo`:

```javascript
findBelongsTo: function(store, record, url, relationship) {
  var related = record[relationship.key];
  // FIXME Without this, it was making unnecessary calls, but cannot create test to verify.
  if(related) { return; }
  return this.ajax(url, 'GET');
}
```

Due to this `FIXME`, `belongsTo` relationships don't get loaded if the key is found on the record.

Example model:

```javascript
export default DS.Model.extend({
    vitals: DS.belongsTo('vitals', { async: true })
});
```

When trying to `get` the `vitals` property, the `if(related) { return; }` block is executed and a request is never fired.

green-arrow commented:
Additional issue that I ran across: specifying an empty array for the root `data` attribute fails.

When `normalizePayload` is called, we recognize that we're dealing with an array and call `extractArrayData`:

```javascript
extractArrayData: function(data, payload) {
  var type = data.length > 0 ? data[0].type : null, serializer = this;
  data.forEach(function(item) {
    if(item.links) {
      serializer.extractRelationships(item.links, item);
      //delete data.links;
    }
  });

  payload[type] = data;
}
```

If no data is present, we punt and say the `type` is null. This causes issues (particularly in the case where you are looking up a `hasMany` relationship). The serializer cannot resolve the type, and we get an error:

```
Error: Assertion Failed: The response from a findHasMany must be an Array, not undefined
```

eneuhauser commented:
@jonkoops, I added 'application/vnd.api+json' in the list of accept headers and set the content type for non-GET requests. Both of these are overridable as properties in the adapter without having to override the whole ajaxOptions method.

@green-arrow, I've fixed the empty array issue. I also removed my override of findBelongsTo. I'm not seeing the extra requests I use to, so some other logic must have fixed that.

jonkoops commented:
@eneuhauser Looks good to me 👍

green-arrow commented:
@eneuhauser awesome.

Additional question. The spec says that you could include a query param `?include=relationship`, making the server sideload the specified relationship. This means that the property on the model will most likely be marked as `{ async: true }`. In this case, calling `model.get('relationship')` results in an API call being made. I would guess that if the record has already been side-loaded, we wouldn't want to kick off another API call.

Thoughts?

eneuhauser commented:
@green-arrow, if you look at the [compound-document-test](https://github.com/eneuhauser/ember-json-api/blob/master/tests/integration/specs/compound-documents-test.js), that's the use case it's validating. The test case ensures that regardless if the relationship is marked as async true or false, if a relationship is returned in the `included` section, it will not make another request.

This is the case when you do `model.get('relationship;)`; however, if you do `store.find('model')`, it will always make a new call.

green-arrow commented:
@eneuhauser gotcha. I will have to do some digging on my end then, because I am seeing more calls executed even though the `belongsTo` relationship has already been loaded in the `included` section.

JoshSmith commented:
@eneuhauser is this supposed to kick off API calls when `hasMany` and `belongsTo` are marked as `{ async: true }`?

green-arrow commented:
@JoshSmith @eneuhauser stated that API calls should not be fired in this case if they've already been loaded via the `included` section, but I am still seeing API calls fired for both relationship types.

eneuhauser commented:
@JoshSmith, the app will fire API calls when calling `get` on a `hasMany` and `belongsTo` that are marked as `{ async: true }` when those resources have not already been pushed to the store. The resources can be pushed to the store by either being returned in an `included` block within the response or from a previous call.

@green-arrow, I'm not sure if you're still having your issue with having extra calls, but I had seen that once before, which is why I use to have the `if(related) { return; }` logic in the findBelongsTo. Unfortunately, I can't explain how that issue got cleared up for me and it doesn't exist in the integration tests. What's even odder is that `var related = record[relationship.key];` should never return a value because `record` is a `DS.Model` and `relationship.key` shouldn't be a property.

green-arrow commented:
@eneuhauser the following prevents duplicate requests:

`app/adapters/application.js`

```javascript
findBelongsTo: function(store, snapshot, url, relationship) {
  if(snapshot.belongsTo(relationship.key)) { return; }

  return this._super(store, snapshot, url, relationship);
}
```

*Note*: This is using `ember-data 1.0.0-beta.16` and taking advantage of the new `snapshot` API. By detecting if the `belongsTo` relationship exists on the snapshot, we can stop another call from being made.

It seems to me, though, that this behavior should be native to ember-data and not something that the `JsonApiAdapter` should have to do. If the record is already loaded, should ember-data simply not execute the `findBelongsTo` method on the adapter?

green-arrow commented:
I think I've narrowed it down a bit more, but now I'm stuck. In ember-data, there is the following code:

```javascript
BelongsToRelationship.prototype.getRecord = function() {
  //TODO(Igor) flushCanonical here once our syncing is not stupid
  if (this.isAsync) {
    var promise;
    if (this.link) {
      var self = this;
      promise = this.findLink().then(function() {
        return self.findRecord();
      });
    } else {
      promise = this.findRecord();
    }

    return PromiseObject.create({
      promise: promise,
      content: this.inverseRecord
    });
  } else {
    Ember.assert("You looked up the '" + this.key + "' relationship on a '" + this.record.constructor.typeKey + "' with id " + this.record.get('id') +  " but some of the associated records were not loaded. Either make sure they are all loaded together with the parent record, or specify that the relationship is async (`DS.belongsTo({ async: true })`)", this.inverseRecord === null || !this.inverseRecord.get('isEmpty'));
    return this.inverseRecord;
  }
};
```

The issue is that `this.link` is populated with a link to the record, so a new promise is being kicked off and a request made. If `this.link` was falsey, a new `PromiseObject` would be created with its content set to `this.inverseRecord`, whose value is the already loaded record.

In `json-api-serializer`, there is a method called `extractRelationships`. Near the bottom of the method, there is this `if` block:

```javascript
if(id) {
  resource[link] = id;
}
```

If I change that block to the following, additional requests are not made:

```javascript
if(id) {
  resource[link] = id;
  delete resource.links[link];
}
```

As far as I can tell, this doesn't seem to have any adverse side effects. I have not pulled down the source and run the tests though. Could someone try what I have above and see if all tests pass? I would do that myself but I have to hop off for a while to attention my kids :smile:

tobscure commented:
Polymorphic relationships are broken with this because the `type` key of any `linkage` objects is diregarded. I was able to fix by replacing the `getLinkageId` and `getLinkageIds` functions at the very bottom of json-api-serializer.js:

```javascript
function getLinkageId(linkage) {
  if(Ember.isEmpty(linkage)) { return null; }
  return (Ember.isArray(linkage)) ? getLinkageIds(linkage) : {type: Ember.String.singularize(linkage.type), id: linkage.id};
}
function getLinkageIds(linkage) {
  if(Ember.isEmpty(linkage)) { return null; }
  var ids = [], index, total;
  for(index=0, total=linkage.length; index<total; ++index) {
    ids.push({type: Ember.String.singularize(linkage[index].type), id: linkage[index].id});
  }
  return ids;
}
```

green-arrow commented:
By adding the `delete` line above, the code that @tobscure provided, and the following to the `adapter`, I think we're pretty much covered:

```javascript
findBelongsTo: function(store, snapshot, url, relationship) {
  if(snapshot.belongsTo(relationship.key)) { return; }

  return this._super(store, snapshot, url, relationship);
},
findHasMany: function(store, snapshot, url, relationship) {
  if(snapshot.hasMany(relationship.key).get('length')) { return []; }

  return this._super(store, snapshot, url, relationship);
}
```

eneuhauser commented:
Unfortunately, none of these pass the tests.

@green-arrow, by definition of the model, `snapshot.belongsTo(relationship.key)` should always return true regardless if it's loaded. This is similar to what I was trying before. Essentially, you need to see if the relationship has already been loaded without calling `get`, since that results in an infinite loop. But as you stated, that really should be the job of ember-data, not the adapter. As for deleting the link, the link is used for other things, like updating the record.

@tobscure, `type` isn't lost because it is defined in the relationship.

If you could send me a sample of your response, I can see if I could duplicate the problem.

green-arrow commented:
@eneuhauser that's not correct. `snapshot.belongsTo(relationship.key)` will return a value if one is already loaded. Otherwise, `undefined` is returned. At least that is what is stated in the ember-data docs:

http://emberjs.com/api/data/classes/DS.Snapshot.html#method_belongsTo

> Calling belongsTo will return a new Snapshot as long as there's any data available, such as an ID. If there's no data available belongsTo will return undefined.

I was not aware that the link was used for other things. In that case, it seems that there is a bug in ember-data in the method I posted above. It seems like `this.inverseRecord` should be the first thing checked instead of `this.link`, and if it is defined, the promise immediately resolved and no AJAX call made. Does that seem right to you?

Maybe @mixonic or somebody else familiar with ember-data could chime in on that?

tobscure commented:
@eneuhauser Given a polymorphic relationship setup like this:

```javascript
App.Activity = DS.Model.extend({
  subject: DS.belongsTo('subject', {polymorphic: true})
});

App.Subject = DS.Model.extend();
App.Post = App.Subject.extend();
App.Discussion = App.Subject.extend();
```

And a JSON-API response like this:

```
{
  "data": [
    {
      "type": "activity",
      "id": "1",
      "links": {
        "subject": {
          "linkage": { "type": "post", "id": "1" }
        }
      }
    },
    {
      "type": "activity",
      "id": "2",
      "links": {
        "subject": {
          "linkage": { "type": "discussion", "id": "2" }
        }
      }
    }
  ],
  "included": [
    {
      "type": "post",
      "id": "1"
    },
    {
      "type": "discussion",
      "id": "1"
    }
  ]
}
```

Currently, this serializer is normalizing the payload to this:

```
{
  "subject": 1
}
```

Which causes Ember Data to freak out because it doesn't know which model type the relationship is to. Instead, my code normalizes it to:

```
{
  "subject": { "type": "post", "id": 1 }
}
```

Which fixes the problem. It was a quick fix; it shouldn't include the type if it's not a polymorphic relationship (and I'm guessing that's the cause of the failed tests).

green-arrow commented:
The code I had for `findHasMany` won't work.

```javascript
findHasMany: function(store, snapshot, url, relationship) {
  if(snapshot.hasMany(relationship.key).get('length')) { return []; }

  return this._super(store, snapshot, url, relationship);
}
```

The next method called is `extractArray`, which just returns `Ember.A()` if no payload is defined.

The root of this problem seems to be the prototypes for `BelongsTo` and `HasMany` executing an ajax call if `this.link` is defined. I have no idea how to resolve this problem.

green-arrow commented:
Sorry for spamming this PR, but I did manage to find a super hacky workaround for `findHasMany`:

```javascript
findHasMany: function(store, snapshot, url, relationship) {
  if(snapshot.hasMany(relationship.key).get('length')) { return new Ember.RSVP.Promise((resolve, reject) => reject()); }

  return this._super(store, snapshot, url, relationship);
}
```

In all, my application adapter looks like this, and I'm not seeing any extra requests being made for data loaded via the `included` section, and all my side-loaded data appears correctly.

```javascript
import Ember from 'ember';
import JsonApiAdapter from 'ember-json-api/json-api-adapter';
import ENV from '../../config/environment';

export default JsonApiAdapter.extend({
  host: ENV.APP.adapter.host,
  namespace: ENV.APP.adapter.namespace,
  findBelongsTo: function(store, snapshot, url, relationship) {
    if(snapshot.belongsTo(relationship.key)) { return; }

    return this._super(store, snapshot, url, relationship);
  },
  findHasMany: function(store, snapshot, url, relationship) {
    if(snapshot.hasMany(relationship.key).get('length')) { return new Ember.RSVP.Promise((resolve, reject) => reject()); }

    return this._super(store, snapshot, url, relationship);
  }
});
```

green-arrow commented:
@tobscure how are you making polymorphic relationships work with this serializer? It seems like polymorphic support was explicitly removed by @eneuhauser 

```javascript
/**
     * Use "links" key, remove support for polymorphic type
     */
    serializeBelongsTo: function(record, json, relationship) {
      var attr = relationship.key;
      var belongsTo = record.belongsTo(attr);
      var type = this.keyForRelationship(relationship.type.typeKey);
      var key = this.keyForRelationship(attr);

      if (isNone(belongsTo)) return;

      json.links = json.links || {};
      json.links[key] = belongsToLink(key, type, get(belongsTo, 'id'));
    }
```

tobscure commented:
@green-arrow I haven't been saving polymorphic relationships, only fetching them

green-arrow commented:
@tobscure what does your payload look like? I have an async polymorphic `belongsTo` that isn't being loaded properly.

tobscure commented:
Pretty much exactly what is in https://github.com/kurko/ember-json-api/pull/71#issuecomment-85333350... mine isn't async though, maybe that could be your problem?

green-arrow commented:
I'm willing to bet that async is the problem. I know other people have had issues with polymorphic + async relationships. 

So, where do we stand on the status of this PR?

green-arrow commented:
For anyone interested in polymorphic async relationships with `ember-json-api`, I managed to get a `belongsTo` relationship working. I modified my `findBelongsTo` method as follows:

```javascript
findBelongsTo: function(store, snapshot, url, relationship) {
  if(snapshot.belongsTo(relationship.key)) { return; }

  return this._super(store, snapshot, url, relationship).then((response) => {
    if(relationship.options.polymorphic) {
      relationship.type = store.modelFor(response.data.type);
    }

    return data;
  });
},
```

By replacing the relationship type with the actual concrete type returned from the API, ember-data is able to correctly serialize the record and push it into the store.

I'm not sure exactly how much of a hack this is, so maybe @eneuhauser can chime in with his thoughts.

Regardless of that, I feel like this PR is at a point where it would be extremely beneficial to merge it.

JoshSmith commented:
I'm trying to use the fix provided by @tobscure but nothing seems to fire off new API calls for non-included links. My API looks to spec (using an RC3 PR for AMS), so I'm not really sure what's going wrong. I can post some more here if useful.

eneuhauser commented:
There currently aren't any polymorphic unit tests. If anyone would like to build one, you can make a pull request against my repository. I've tried all the proposed fixes in here, but they break other unit tests.

@green-arrow, the guard statement `if(snapshot.belongsTo(relationship.key)) { return; }` in `findBelongsTo` breaks the tests. Even when it returns `[]`. If I remove the guard statement, the rest of the code does not break the tests, but I don't know if that fixes your problem.

Also, when I upgrade to v1.0.0-beta.16, I get the following warning `DEPRECATION: Usage of 'snapshot.constructor' is deprecated, use 'snapshot.type' instead.`. I'd like to fix those occurrences before upgrading.

@tobscure, thanks for the insights on polymorphic. Do you know of a way to get the relationship type at the time of `extractRelationships` so we can return the proper response?

I don't have a lot of time this week, but I'll do what I can. Instead, I have to context shift and work on an Angular project.

green-arrow commented:
@JoshSmith I just submitted a PR that doesn't fire off requests for data loaded via an `included` section, but will fire off requests for relationships that have not yet been loaded. All tests pass for me. As soon as @eneuhauser gets the time, I'm hoping that he can take a look and merge it.

kurko commented on CHANGELOG.md:
```diff
@@ -1,5 +1,9 @@
 
+### 0.4.0
+
+* updating to RC2 standard of JSON API
```
You mean RC3 :smile:

eneuhauser replied:
Not sure if it matters, but 4.0 was RC2, 4.1 was RC2.1? where linked was changed to included and resource became related. 4.1 is somewhat backwards compatible in that you could override what those properties are called, but 4.2 is where it became RC3, which includes the consistent linkage format. If you don't care about the history, I can just call the current version 4.0 and update the changelog accordingly.

kurko commented on dist/ember-json-api.js:
```diff
@@ -123,24 +149,10 @@ define("json-api-adapter",
       pathForType: function(type) {
         var decamelized = Ember.String.decamelize(type);
-        return Ember.String.pluralize(decamelized);
+        return Ember.String.pluralize(decamelized).replace(/_/g, '-');
```
The spec is not specific about `-`, so instead of enforcing it right now, may I suggest adding to the README a piece on how to override `pathForType` to change it from `-` to snakecase.

Personally, I'd leave it as snakecase (because AMS, one of the biggest players in the serializers-land uses snakecase) and inform in the README how to change it to `-`, but that's your choice :smile:

kurko commented on dist/ember-json-api.js:
```diff
@@ -319,9 +373,9 @@ define("json-api-adapter",
-        var key = this.keyForRelationship(attr);
+        var attr = relationship.key,
+          type = this.keyForRelationship(relationship.type.typeKey),
+          key = this.keyForRelationship(attr);
```
Do you mind not changing the style in this PR? Nothing against the change, but I'd rather do it in a separate PR so that we can discuss it specifically.

Using `var` in each line makes it easier to debug stuff.

eneuhauser replied:
I used this style extensively, as it's a best practice from [Effective JavaScript](http://effectivejs.com/). That said, I agree that it could difficult to debug sometimes.

kurko commented:
This PR looks amazing. A few rough edge, but amazing nonetheless. I'd like to have someone testing in their apps before merging this. Please let us know about the problems or successes you're having :smile:  (if you retweet https://twitter.com/kurko/status/580735904999952384 we might get more visibility)

sirvine commented:
@kurko The only problem I've had so far is that I had `coalesceFindRequests` turned on initially after converting from ActiveModelAdapter. The PR works without that option, and I think it makes sense to get this PR into the addon, and deal with coalescing queries later.

JoshSmith commented:
This is still not firing any API requests for me.

My JSON API response:

```
{
  data: {
    id: "3",
    title: "Chopped Salad",
    awaiting_answers: false,
    displayable: true,
    cover_photo_public_id: "u1gctflzpczfilyu8vph",
    plural: false,
    cook_time: null,
    paid: true,
    trialing: false,
    accessible: false,
    type: "recipes",
    links: {
      steps: {
        linkage: [
          {
            type: "steps",
            id: "64"
          },
          {
            type: "steps",
            id: "66"
          },
          {
            type: "steps",
            id: "67"
          }
        ]
      },
    }
  }
}
```

What I had expected to see was an API request fired for `GET /steps/67`, for example. (And eventually it'd be nice if those were coalesced a la @sirvine's `coalesceFindRequests`). But I don't see any other API requests after `GET /recipes/3`.

Am I the culprit here?

JoshSmith commented:
Bump.

fotinakis commented:
+1 for getting this PR in!

I've been able to successfully upgrade to this and run from eneuhauser's fork, albeit finding out that you must upgrade to ember-data 1.0.0-beta.16 (beta.15 is incompatible).

green-arrow commented:
@JoshSmith I'm not sure why you're having this issue. One of the tests explicitly tests for a scenario where additional resources are requests if provided in `linkage`:

`tests/specs/multiple-resource-links-test`:

```javascript
var get = Ember.get, set = Ember.set;
var env;
var responses, fakeServer;

module('integration/specs/multiple-resource-links-test', {
    setup: function() {
        fakeServer = stubServer();

        responses = {
            posts_not_compound: {
                data: [{
                    type: 'posts',
                    id: '1',
                    title: 'Rails is Omakase',
                    links: {
                        author: {
                            related: '/posts/1/author',
                            linkage: {
                                type: 'authors',
                                id: '2'
                            }
                        }
                    }
                }, {
                    type: 'posts',
                    id: '2',
                    title: 'TDD Is Dead lol',
                    links: {
                        author: {
                            related: '/posts/2/author',
                            linkage: {
                                type: 'authors',
                                id: '1'
                            }
                        }
                    }
                }]
            },
            post_1_author: {
                data: {
                    type: 'authors',
                    id: '2',
                    name: 'dhh'
                }
            },
            post_2_author: {
                data: {
                    type: 'authors',
                    id: '1',
                    name: 'ado'
                }
            }
        };

        env = setupStore(setModels());
        env.store.modelFor('post');
        env.store.modelFor('comment');
        env.store.modelFor('author');
    },

    teardown: function() {
        Ember.run(env.store, 'destroy');
        shutdownFakeServer(fakeServer);
    }
});

asyncTest('GET /posts/1 calls later GET /posts/1/comments when Posts has async comments', function() {
    var models = setModels({
        authorAsync: true
    });
    env = setupStore(models);

    fakeServer.get('/posts', responses.posts_not_compound);
    fakeServer.get('/posts/1/author', responses.post_1_author);
    fakeServer.get('/posts/2/author', responses.post_2_author);

    Em.run(function() {
        env.store.find('post').then(function(records) {
            equal(records.get('length'), 2, 'there are 2 posts');

             var post1 = records.objectAt(0);
             var post2 = records.objectAt(1);
             var promises = [];

             promises.push(post1.get('author').then(function(author) {
                equal(author.get('name'), 'dhh', 'post1 author');
             }));
             promises.push(post2.get('author').then(function(author) {
                 equal(author.get('name'), 'ado', 'post2 author');
             }));

             Ember.RSVP.all(promises).then(start);
        });
    });
});
```

JoshSmith commented:
@green-arrow I'm not really sure either. Happy to pair with someone or provide more info where needed, but I don't see what could be the problem here. Tried debugging but got nowhere.

JoshSmith commented:
Do I need to very specifically call `.get` to have the request happen? I'm going from a working app operating in the wild under `ActiveModelAdapter` for some time to using the `JsonApiAdapter` instead. I would expect the functionality I had to not break so severely.

JoshSmith commented:
I also notice that there's a `related` key that I don't have in my API's JSON response, but wouldn't think that would make a difference.

JoshSmith commented:
The way I'm getting steps is like so:

```
var RecipeController = Ember.ObjectController.extend({
  steps: (function() {
    return Ember.ArrayProxy.createWithMixins(Ember.SortableMixin, {
      sortProperties: ['position'],
      content: this.get('content.steps')
    });
  }).property('content.steps.[]'),
});
```

green-arrow commented:
I'm not sure if this will make a difference, but do you need to use an
`ObjectController` here? I believe that and `ArrayController` will be
deprecated in the future.

I assume you have a model backing this controller, so I'd change
`content.steps` to `model.steps`.

I'm also not familiar with `content.steps.[]`. If you're wanting to base
the CP off of each item in the `steps` array, I think the proper syntax is
`model.steps.@each`.

Again, I have no idea if any of these things will help. My first suggestion
would be to use a regular `Controller` instead of `ObjectController` and
see if that helps any.

I don't think the `related` link should have anything to do with it.
On Fri, Mar 27, 2015 at 9:21 PM Josh Smith <notifications@github.com> wrote:

> The way I'm getting steps is like so:
>
> var RecipeController = Ember.ObjectController.extend({
>   steps: (function() {
>     return Ember.ArrayProxy.createWithMixins(Ember.SortableMixin, {
>       sortProperties: ['position'],
>       content: this.get('content.steps')
>     });
>   }).property('content.steps.[]'),
> });
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kurko/ember-json-api/pull/71#issuecomment-87137173>.
>

green-arrow commented:
I'm not sure if this will make a difference, but do you need to use an
`ObjectController` here? I believe that and `ArrayController` will be
deprecated in the future.

I assume you have a model backing this controller, so I'd change
`content.steps` to `model.steps`.

I'm also not familiar with `content.steps.[]`. If you're wanting to base
the CP off of each item in the `steps` array, I think the proper syntax is
`model.steps.@each`.

Again, I have no idea if any of these things will help. My first suggestion
would be to use a regular `Controller` instead of `ObjectController` and
see if that helps any.

I don't think the `related` link should have anything to do with it.

```javascript
var RecipeController = Ember.Controller.extend({
  steps: Ember.computed('model.steps.@each', function() {
    return Ember.ArrayProxy.createWithMixins(Ember.SortableMixin, {
      sortProperties: ['position'],
      content: this.get('model.steps')
    });
  })
});
```

JoshSmith commented:
I thought the proper syntax was `.[]` for an array. I always get confused by the salient difference between that and `.@each`.

I updated this (aside from the `ObjectController` to `Controller` change) and still nothing. No request gets fired.

green-arrow commented:
Would you be open to trying the change to a normal `Ember.Controller` and `content` => `model`?

Aside from that, can you verify that you're actually getting inside of the `steps` computed property? If so, what happens if you just return `this.get('model.steps')`?

Is your code private, or is it in a public place where someone can take a look?

sirvine commented:
@JoshSmith Are you calling get on something else to pull the `recipe` model? If so, you might need to chain the `steps` get call after the `recipe` promise resolves.

JoshSmith commented:
If I manually call the `model.get('steps')` in my `afterModel` hook then the API requests are successfully made.

Is this expected behavior? It was non-obvious to me that I would need to manually call `.get` on every single relationship in order to load related data.

Unfortunately @green-arrow the code is in a private repo.

green-arrow commented:
@JoshSmith if you're relationship is marked as `async`, it is not fetched until it is actually referenced (either in a template or via a `get` call).

JoshSmith commented:
But the `this.get('model.steps')` I tried inside the `RecipeController` I would imagine would have hit that, and that is definitely hit by my templates. I even commented that out and just tried to iterate over `steps` in the template, which also does not work.

eneuhauser commented:
For those using my branch, please checkout my latest changes. I've merged the changes that have been checked into the master since my fork. [One](https://github.com/kurko/ember-json-api/pull/65) in the master and [one](https://github.com/eneuhauser/ember-json-api/pull/2) into my branch improve polymorphic relationships, so hopefully that will resolve the major outstanding issues.

I believe my last set of changes already required Ember Data to 1.0 beta-16, but these latest changes definitely require an upgrade.

green-arrow commented:
@eneuhauser I pulled down the latest changes, but still have issues with polymorphic relationships. Unfortunately, the model is not being pushed onto my store. In order to get that to work, I need the following extension of `findBelongsTo` in my adapter:

```javascript
findBelongsTo: function(store, snapshot, url, relationship) {
  var promise = this._super(store, snapshot, url, relationship);

  if(promise) {
    promise = promise.then((response) => {
      if(relationship.options.polymorphic) {
        relationship.type = store.modelFor(response.data.type);
      }

      return response;
    });
  }

  return promise;
}
```

fotinakis commented:
Now that compound documents will be fully supported, is there way to add query params for the `include` key when fetching a single record?

I was thinking something like:
```javascript
model: function(params) {
  return this.store.find('post', params.id, {include: 'comments'})
}
```
...but this doesn't work, it seems the store interface doesn't support single-record query params at all yet and it's a 2-year old issue (https://github.com/emberjs/data/issues/1576).

Any workarounds come to mind? I think this will come up more and more as people start using JSON API more since `include` is a powerful feature there.

tobscure commented:
Implement your own findQueryOne method, based on this PR https://github.com/emberjs/data/pull/2584

initializers/find-query-one.js:
```javascript
import DS from 'ember-data';

export default {
  name: 'find-query-one',
  initialize: function(container) {
    DS.Store.reopen({
      findQueryOne: function(type, id, query) {
        var store = this;
        var typeClass = store.modelFor(type);
        var adapter = store.adapterFor(typeClass);
        var serializer = store.serializerFor(typeClass);
        var url = adapter.buildURL(type, id);
        var ajaxPromise = adapter.ajax(url, 'GET', { data: query });

        return ajaxPromise.then(function(rawPayload) {
          var extractedPayload = serializer.extract(store, typeClass, rawPayload, id, 'find');
          return store.push(typeClass, extractedPayload);
        });
      }
    });
  }
};
```

fotinakis commented:
I'll check it out, thanks @tobscure.

green-arrow commented:
@eneuhauser looks like the code I put in place for the `findHasMany` duplicate requests fix actually causes some issues. When executing `model.reload()` or `model.get('relationship').reload()`, that method is hit once again. However, since it was already previously loaded, no API call is executed.

I'll try to look into that some more this week, but I welcome any thoughts / ideas.

green-arrow commented:
@kurko where do we stand on getting this merged? It would be nice to be able to get that done and then be able to open up separate issues for the things we run into (like the model reloading things I just mentioned above). Thanks!

fotinakis commented:
Definite :+1: for getting this merged in and filing followup bugs, I have some small issues I've been holding on filing until this gets merged.

jamonholmgren commented:
So far it's working well for me! :+1:

kurko commented:
@green-arrow I'm dying to merge this, but I don't want bugs to be introduced in it. If the issues you have were present before, then it's fine. I don't have a RC3 app right now to test it myself, so I need you folks to tell me whether the pending issues are blocking or not.

> @eneuhauser looks like the code I put in place for the findHasMany duplicate requests fix actually causes some issues

How do we stand with this?

green-arrow commented:
@kurko the issue it presents is that the `findMany` relationship is never reloaded, even if `model.get('relationship').reload()` is executed. Since the relationship was loaded already, it never executes another API request. 

I have not yet found a way to make this method re-execute the request if a `reload` has been requested.

eneuhauser commented:
This conversation is getting huge! I have enabled [issues within my repository](https://github.com/eneuhauser/ember-json-api/issues) to better track outstanding items. I have been wrapped up in a couple of projects, but I should have time soon to address reported issues.

fotinakis commented:
@kurko we need you! :) You good with merging this?

dopudesign commented:
Thank you for all the work on this! This is exactly the solution I've been looking for, for the errors I've been running into :smile:

jonkoops commented:
@eneuhauser Seems like issue #3 is a blocking issue in specification compliance.

andrewbranch commented:
@jonkoops I submitted a PR to @eneuhauser implementing the PUT/PATCH swap needed for specification compliance: eneuhauser/ember-json-api#6

kurko commented:
Ok, so unless someone comes shouting and saying this is not working, I'm merging this tonight and release a new version :smile:

fotinakis commented:
:+1: :+1: :+1: Yay!

jamonholmgren commented:
Awesome!

eneuhauser commented:
Awesome, thanks! There are [3 issues](https://github.com/eneuhauser/ember-json-api/issues) reported in my repository that I haven't had a chance to look at yet. Additionally, I'm [working on the findHasMany issue](https://github.com/eneuhauser/ember-json-api/tree/has-many). I'm hoping to have time tomorrow to work on them. If you merge the current pull request, I can submit separate pull requests to solve those issues.

jamonholmgren commented:
@eneuhauser Thanks for your hard work on this. And @kurko of course. As an open source project maintainer myself, I know these hours sometimes go unnoticed. Much appreciated.

jonkoops commented:
I think we should get eneuhauser#8 fixed first, it's a minor issue but it could cause some confusion.

eneuhauser commented:
I've made some comments in [issue#8](https://github.com/eneuhauser/ember-json-api/issues/8). Would love some feedback if I should merge my fix.

joelcox commented:
Thanks for your great work on this @eneuhauser. We've been using your branch for some time now and I haven't run into any big issues so far.

jonkoops commented:
@kurko In my opinion we should get this merged, @eneuhauser anything blocking in your opinion?

eneuhauser commented:
@jonkoops and @kurko, I think RC3 is good to go. There will be more changes before 1.0 hits. Someone already opened [Issue 10](https://github.com/eneuhauser/ember-json-api/issues/10) about a change since RC3.

Since ember-json-api is one of the de facto implementations of JSON API, I think it would be a good idea to keep up with the release candidates. I can begin work on that issue, but until another release candidate hits or 1.0, I'm not going to merge it in.

Boubalou commented:
@eneuhauser Based on the latest RC3 spec, there is an attributes field that hold all of the model fields, is there any plan on updating this PR to support the final revision RC3?

green-arrow commented:
:+1: on getting this merged

jonkoops commented:
@kurko https://youtu.be/vCadcBR95oU

kurko commented:
Folks, I'm merging this soon. I'm on vacation and without computer. Let me
see if I can do something via mobile this weekend 😄
On Tue, May 12, 2015 at 7:09 PM Jon Koops <notifications@github.com> wrote:

> @kurko <https://github.com/kurko> https://youtu.be/vCadcBR95oU
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kurko/ember-json-api/pull/71#issuecomment-101352784>.
>

kurko commented:
:smile: 

Now, would someone take the torch for RC4? Seems pretty straightforward

eneuhauser commented:
Thanks for merging! I have been heads down on a few non-Ember projects, but I am days away from a vacation. I should have some downtime to update to RC4. Someone has already submitted a pull request into my repo. I'll also look into upgrading Ember and Ember Data versions. According to the [status page](http://jsonapi.org/#update-history), JSON API 1.0 is slated for tomorrow. That will get some consistencies in implementations to get everyone on the same page.

josepjaume commented:
For the record: The latest developments are in https://github.com/eneuhauser/ember-json-api/pull/14. I'm using it in my app and seems to work fine :+1:
